### PR TITLE
Add Metric Benchmark tests to integration tests

### PR DESCRIPTION
**Pull request by TheRasanjana**
Benchmark test are added for the metrices No. of SQLs executed and No. of external API calls for the following scenarios.
1. Creating an API
2. Publishing an API
3. Retrieving all APIs from publisher (page 1)
4. Retrieving all APIs from store (page 1)
5. Retrieving an API from publisher
6. Retrieving an API from store
7. Creating an Application
8. Subscribing to an API
9. Generating an access token - JWT
10. Invoking the created API with JWT

- As the logs are read from correlation logs, tests were implemented with restassuared to add the required header values for mapping.
- Benchmark values are defined in "src/test/resources/benchmark-values"
- For each test scenario relevant logs will be extracted and can be found in "target/logs/benchmark-tests"

### CODE REVIEW

File: modules/integration/tests-integration/tests-backend/pom.xml [hunk 1]
```diff
@@ -759,5 +759,10 @@ under the License.
             <artifactId>nimbus-jose-jwt</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>io.rest-assured</groupId>
+            <artifactId>rest-assured</artifactId>
+            <version>4.3.1</version>
```
Comment (bhathiya) [inline modules/integration/tests-integration/tests-backend/pom.xml:765]: please add test scope

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/benchmarktest/BenchmarkTestCase.java [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+    public static Object[][] DataProvider() {
+
+        return new Object[][]{
+                {"jdbc"}, {"http"}
+        };
+    }
+
+    @BeforeClass(alwaysRun = true)
+    public void setEnvironment() throws Exception {
+// Benchmark values are defined in "src/test/resources/benchmark-values"
```
Comment (bhathiya) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/benchmarktest/BenchmarkTestCase.java:61]:
```suggestion
    // Benchmark values are defined in "src/test/resources/benchmark-values"
```

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/benchmarktest/BenchmarkTestCase.java [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+
+        List<String> apiIdList = new ArrayList<>();
+        int noOfApisCreated = 20;
+        int noOfAPISRetrieved = 10;
+        benchmark = BenchmarkUtils.getBenchmark(testMetric, "RETRIEVE_ALL_STORE");
+        testName = method.getName();
+        context = "testcontext_" + testName;
+        String corellationID = benchmarkUtils.setActivityID();
+        for (int i = 0; i < noOfApisCreated; i++) {
+            apiUUID = benchmarkUtils.createRestAPI(testName + testMetric + i, context + testMetric + i, "");
```
Comment (bhathiya) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/benchmarktest/BenchmarkTestCase.java:141]: why don't we reuse the common data created? We need to make sure the tests run with the minimum amount of time possible.

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/benchmarktest/BenchmarkTestCase.java [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+            benchmarkUtils.publishAPI(apiUUID, "");
+            apiIdList.add(apiUUID);
+        }
+        while (!(benchmarkUtils.getDevPortalApiCount() >= noOfApisCreated)) {
+            Thread.sleep(1000);
```
Comment (bhathiya) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/benchmarktest/BenchmarkTestCase.java:146]: we need to break this after a given number of retries. otherwise, this can stuck forever.

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/benchmarktest/BenchmarkUtils.java [hunk 5]
```diff
@@ -0,0 +1,551 @@
...
+        int noOfLinesExecuted = 0;
+        try (BufferedReader bufferedReader = new BufferedReader(new FileReader(IN_FILE_PATH));
+//             lines executed are logged in to a file starting with test method name,
+             Writer writer = new BufferedWriter(
+                     new OutputStreamWriter(new FileOutputStream(OUT_FILE_PATH + logFile + "_" + testType + ".log"), "utf-8"))
+        ) {
+            String readLine;
+            while ((readLine = bufferedReader.readLine()) != null) {
+                String logLine = readLine.toLowerCase();
+//                 lines executed will be captured with log Attribute
```
Comment (bhathiya) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/benchmarktest/BenchmarkUtils.java:108]: align all comments with code.

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/benchmarktest/BenchmarkUtils.java [hunk 6]
```diff
@@ -0,0 +1,551 @@
...
+                if (logLine.contains("|" + logAttribute + "|")) {
+                    String regex = "(\\d{2}:\\d{2}:\\d{2})";
+                    Matcher m = Pattern.compile(regex).matcher(logLine);
```
Comment (bhathiya) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/benchmarktest/BenchmarkUtils.java:111]:
```suggestion
                    Matcher matcher = Pattern.compile(regex).matcher(logLine);
```

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/benchmarktest/BenchmarkTestCase.java (hunk 4 above)
Comment (bhathiya) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/benchmarktest/BenchmarkTestCase.java:146]: make sure to do this everywhere relevant.

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/benchmarktest/BenchmarkUtils.java [hunk 7]
```diff
@@ -0,0 +1,551 @@
...
+        if (!f.exists()) {
+            f.createNewFile();
+        }
+        BufferedWriter bw = new BufferedWriter(new FileWriter(f, true));
+        bw.append(testName + " ===== Actual count is " + actual + "   Benchmark is  " + benchmark);
+        bw.newLine();
+        bw.close();
+    }
+
+    public void generateConsumerCredentialsAndAccessToken() throws IOException {
```
Comment (bhathiya) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/benchmarktest/BenchmarkUtils.java:198]: why do we need non-static methods in a util class?

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/benchmarktest/BenchmarkUtils.java [hunk 8]
```diff
@@ -0,0 +1,551 @@
...
+                        .formParam("grant_type", "password")
+                        .formParam("username", SUPER_TENANT_USERNAME)
+                        .formParam("password", SUPER_TENANT_PASSWORD)
+                        .formParam("scope", scope).
+                        when().
+                        post(gateway_Url + "/token");
+        String responseBody = response.getBody().asString();
+        JsonPath jsonResponse = new JsonPath(responseBody);
+        statusCode = response.getStatusCode();
+//            Assert.assertEquals(statusCode /*actual value*/,  HttpStatus.SC_OK /*expected value*/, "Incorrect status code returned");
```
Comment (bhathiya) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/benchmarktest/BenchmarkUtils.java:243]: remove commented code

File: modules/integration/tests-integration/tests-backend/src/test/resources/testng.xml [hunk 9]
```diff
@@ -344,4 +344,10 @@
             <class name="org.wso2.am.integration.tests.workflow.WorkflowApprovalExecutorTest"/>
         </classes>
     </test>
+
+    <test name="apim-integration-tests-benchmark" preserve-order="true" parallel="false" group-by-instances="true">
+        <classes>
+            <class name="org.wso2.am.integration.tests.benchmarktest.BenchmarkTestCase"/>
```
Comment (bhathiya) [inline modules/integration/tests-integration/tests-backend/src/test/resources/testng.xml:350]: how long does this take to run completely?

**Review by bhathiya — CHANGES_REQUESTED** [review]
Added inline comments

File: modules/integration/tests-integration/tests-backend/pom.xml (hunk 1 above)
Comment (TheRasanjana) [reply modules/integration/tests-integration/tests-backend/pom.xml:765]: Done

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/benchmarktest/BenchmarkTestCase.java (hunk 3 above)
Comment (TheRasanjana) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/benchmarktest/BenchmarkTestCase.java:146]: Generating common test data before all tests will impact the result counts of other tests.

File: modules/integration/tests-integration/tests-backend/src/test/resources/testng.xml (hunk 9 above)
Comment (TheRasanjana) [reply modules/integration/tests-integration/tests-backend/src/test/resources/testng.xml:350]: About 15 minutes, including the tenant iterations.

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/benchmarktest/BenchmarkTestCase.java (hunk 2 above)
Comment (TheRasanjana) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/benchmarktest/BenchmarkTestCase.java:67]: Done

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/benchmarktest/BenchmarkTestCase.java (hunk 4 above)
Comment (TheRasanjana) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/benchmarktest/BenchmarkTestCase.java:151]: Done

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/benchmarktest/BenchmarkUtils.java (hunk 5 above)
Comment (TheRasanjana) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/benchmarktest/BenchmarkUtils.java:115]: Done

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/benchmarktest/BenchmarkUtils.java (hunk 6 above)
Comment (TheRasanjana) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/benchmarktest/BenchmarkUtils.java:118]: Done

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/benchmarktest/BenchmarkUtils.java (hunk 7 above)
Comment (TheRasanjana) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/benchmarktest/BenchmarkUtils.java]: they should be static. made the changes.

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/benchmarktest/BenchmarkUtils.java (hunk 8 above)
Comment (TheRasanjana) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/benchmarktest/BenchmarkUtils.java:272]: Done

File: modules/integration/tests-integration/tests-backend/src/test/resources/testng.xml (hunk 9 above)
Comment (bhathiya) [reply modules/integration/tests-integration/tests-backend/src/test/resources/testng.xml:350]: we need to reduce this. let's discuss more at the code review.

**Comment by tharindu1st** [comment]
@TheRasanjana please check for github checks failure
